controllers/resources_test.go

```diff
@@ -268,4 +280,165 @@ var _ = Describe("Resource creation and cleanup tests", func() {
+		)
+
+		BeforeEach(func() {
+			ctx = context.Background()
+			a = makeTestRolloutManager()
+			r = makeTestReconciler(a)
+			err := createNamespace(r, a.Namespace)
+			Expect(err).ToNot(HaveOccurred())
+			req = reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      a.Name,
+					Namespace: a.Namespace,
+				},
+			}
+		})
+
+		It("Verify whether RolloutManager creating ServiceMonitor", func() {
+			smCRD, existingSvc := serviceAndServiceMonitorCRD(req.Namespace)
+			Expect(r.Client.Create(ctx, smCRD)).To(Succeed())
+			Expect(r.Client.Create(ctx, existingSvc)).To(Succeed())
+
+			res, err := r.Reconcile(ctx, req)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(res.Requeue).Should(BeFalse(), "reconcile should not requeue request")
+
+			expectedServiceMonitor := serviceMonitor()
+
+			sm := &monitoringv1.ServiceMonitor{}
+			Expect(r.Client.Get(ctx, types.NamespacedName{
+				Name:      DefaultArgoRolloutsMetricsServiceName,
+				Namespace: testNamespace,
+			}, sm)).To(Succeed())
+
+			Expect(sm.Name).To(Equal(expectedServiceMonitor.Name))
+			Expect(sm.Namespace).To(Equal(expectedServiceMonitor.Namespace))
+			Expect(sm.Spec).To(Equal(expectedServiceMonitor.Spec))
+		})
+
+		It("Verify if ServiceMonitor exists, but has different content than we expect then it should update ServiceMonitor", func() {
+			smCRD, existingSvc := serviceAndServiceMonitorCRD(req.Namespace)
+			Expect(r.Client.Create(ctx, smCRD)).To(Succeed())
+			Expect(r.Client.Create(ctx, existingSvc)).To(Succeed())
+
+			existingServiceMonitor := &monitoringv1.ServiceMonitor{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      DefaultArgoRolloutsMetricsServiceName,
+					Namespace: testNamespace,
+				},
+				Spec: monitoringv1.ServiceMonitorSpec{
+					Selector: metav1.LabelSelector{
+						MatchLabels: map[string]string{
+							"app.kubernetes.io/name": "test-label",
+						},
+					},
+					Endpoints: []monitoringv1.Endpoint{
+						{
+							Port: "metrics-test",
+						},
+					},
+				},
+			}
+
+			Expect(r.Client.Create(ctx, existingServiceMonitor)).To(Succeed())
+
+			res, err := r.Reconcile(ctx, req)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(res.Requeue).Should(BeFalse(), "reconcile should not requeue request")
+
+			expectedSM := serviceMonitor()
+
+			Expect(r.Client.Get(ctx, types.NamespacedName{
+				Name:      DefaultArgoRolloutsMetricsServiceName,
+				Namespace: testNamespace,
+			}, existingServiceMonitor)).To(Succeed())
+
+			Expect(existingServiceMonitor.Name).To(Equal(expectedSM.Name))
+			Expect(existingServiceMonitor.Namespace).To(Equal(expectedSM.Namespace))
+			Expect(existingServiceMonitor.Spec).To(Equal(expectedSM.Spec))
+
+		})
+
+		It("Verify ServiceMonitor is not created if the CRD does not exist.", func() {
+			_, existingSvc := serviceAndServiceMonitorCRD(req.Namespace)
+			Expect(r.Client.Create(ctx, existingSvc)).To(Succeed())
+
+			res, err := r.Reconcile(ctx, req)
+			Expect(err).To(HaveOccurred())
```

Comment:
Based on my request for a logic change above, this should no longer return an error.